### ISSUE TITLE
Querying by ID does not work for observe and observeQuery

### ISSUE THREAD

**Issue by Jordan-Nelson** [aws-amplify/amplify-flutter#1594]
### Description

The following subscription on DataStore.observe will not be hit.

```dart
final blog = Blog(name: 'my blog');

Amplify.DataStore.observe(Blog.classType, where: Blog.ID.eq(blog.id)).listen(print);

while (true) {
  await Future.delayed(Duration(milliseconds: 100));
  await Amplify.DataStore.save(blog);
}
```

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [X] DataStore
- [ ] Storage

### Steps to Reproduce

See description

### Platforms

- [ ] iOS
- [ ] Android

### Environment

```bash
N/A
```

### Dependencies

```bash
N/A
```

### Device

N/A

### OS

N/A

### CLI Version

N/A

### Additional Context

The `evaluate` method in `QueryPredicateOperation` used the `field` to find the value to compare in the query predicate. The `field` on `Blog.ID.equals()` is `"Blog.id"`. The evaluate method expect it to be `"id"`.

**Comment by HuiSF** [comment]
This issue sounds similar to #1565

**Comment by Jordan-Nelson** [comment]
> This issue sounds similar to #1565

Interesting. It does look similar. I think the fix will be different, although they might sort of come from the same root issue, which is that the fieldName has the model name prepended for IDs.

When I wrote the in memory query predicate logic (used in observe/observeQuery) I had assumed the `fieldName` was always just the same as the key in the `toJson()` result. This is not true for IDs though since the model name is prepended.

For example, here are the query predicates and the toJson method for a Blog type. The key in `toJson()` is `"id"` but the fieldName in the query predicate is `"blog.id"`

```dart
Map<String, dynamic> toJson() => {
      'id': id,
      'name': _name,
      'createdAt': _createdAt?.format(),
      'file': _file?.toJson(),
      'files': _files?.map((S3Object? e) => e?.toJson()).toList(),
      'posts': _posts?.map((Post? e) => e?.toJson()).toList(),
      'updatedAt': _updatedAt?.format()
    };

static final QueryField ID = QueryField(fieldName: "blog.id");
static final QueryField NAME = QueryField(fieldName: "name");
static final QueryField CREATEDAT = QueryField(fieldName: "createdAt");
static final QueryField FILE = QueryField(fieldName: "file");
static final QueryField FILES = QueryField(fieldName: "files");
static final QueryField POSTS = QueryField(
  fieldName: "posts",
  fieldType: ModelFieldType(
    ModelFieldTypeEnum.model,
    ofModelName: (Post).toString(),
  ),
);
```

Here are three options for fixing this issue. The last two require codegen changes.

### Option 1

The easiest way to resolve this issue would be to ignore everything before the `.` if the fieldName contains a `.`. Basically to just update line 73 below to be `var value = model.toJson()[field.split('.').last];`

https://github.com/aws-amplify/amplify-flutter/blob/080685ac8037ad60b4a4f9678eb9c4a81a4fa30e/packages/amplify_core/lib/src/types/query/query_predicate.dart#L70-L75

I think that would likely work, although it feels like a bit of a hack and I am unsure if there could be edge cases I am not accounting for.

### Option 2

Another possible option would be to add a new method that creates a map with keys that match identically. Using `toJson()` was a hack to get around the fact that Dart doesn't have reflection. Serializing the entire object to json is actually unnecessary and inefficient. What would be ideal is something like this:

```dart
Map<String, dynamic> toFieldNameMap() => {
      'blog.id': id,
      'name': _name,
      'createdAt': _createdAt,
      'file': _file,
      'files': _files,
      'posts': _posts,
      'updatedAt': _updatedAt
    };
```

This involves a codegen change, but it would be a fairly simple change that would require customers immediately upgrade to the latest version of amplify-flutter.

### Option 3

This is IMO the cleanest solution, but it is a larger change and would require that customers upgrade to the latest version of amplify-flutter when they upgrade the CLI.

Instead of creating a method to create a map of the whole object, `QueryField` could have a method to get the field value so that it can be used in comparisons. I actually implemented this idea in a POC. You can see that change [here](https://github.com/aws-amplify/amplify-flutter/pull/1240/files#diff-b2d9c3ad078ebf59a9153c0dec7a8a1c2f8cd903b5eaa936eac982685dfc8174).

A `QueryField` for the ID field on a blog would look something like this:

```dart
static final QueryField<Blog, String> ID = QueryField(
  fieldName: "blog.id",
  getValue: (model) => model.id,
);
```

And the `evaluate` method then looks like:

```dart
bool evaluate(M model) {
  var value = queryFieldOperator.getValue(model);
  return queryFieldOperator.evaluate(value);
}
```

**Comment by Jordan-Nelson** [comment]
Here is a draft of what option 1 would look like: https://github.com/aws-amplify/amplify-flutter/pull/1597

**Comment by HuiSF** [comment]
The fix for this issue has been released with version [0.5.1](https://github.com/aws-amplify/amplify-flutter/releases/tag/v0.5.1).
